### ISSUE TITLE
combine with sphereface

### ISSUE THREAD

**Issue by xizi** [happynear/NormFace#21]
@happynear hi, how can i combine the normface with sphereface, could you give me some help?

**Comment by happynear** [comment]
Simply combine normface with sphereface (normalize both of the feature and weights) will only make the performance worse. I am preparing a paper to describe the correct way to do this. You may see it in two months...
At present, you may try the sphereface algorithm to get more familiar with the parameter tuning.

**Comment by xizi** [comment]
Great, i'm looking forward to your paper and can you remind me when paper is completed?  My wechat name is xizi_fish and we can add as a friend.

**Comment by KaleidoZhouYN** [comment]
@happynear so do you have any idea about coco_loss(present by sensetime,normalized both feature and weghts) and reach an unbelievable high accuracy on LFW only using CASIA-Webface.which means they have solved eyeglasses,expression and large-angle face pose.I'm very confused about whethre this can be solved only use CAISA-Webface.

**Comment by happynear** [comment]
They changed their description in the newest version. They actually used MS-Celeb as the training set.

**Comment by KaleidoZhouYN** [comment]
@happynear  and may I ask how is the progress going for combining normface with sphereface?

**Comment by happynear** [comment]
That's failed. The new algorithm only works on LFW BLUFR protocol. On megaface, it's performance is similar with sphereface. It is not good enough for top conferences.

**Comment by KaleidoZhouYN** [comment]
@happynear  sorry to hear about that,but have you considered the reason why you combine normface and sphereface,is there a theoretical idea that combine normface and sphereface should work?Maybe it's the other reasons(like alignment) that cause it to fail on megaface.

**Comment by happynear** [comment]
Well, normface actually doesn't have a theoretical basis. It is based on a methodology that we should make training and testing consistent. A methodology is not as strong as a mathematical theory. Whether we should normalize the feature or not is still doubtful.

Now I think feature normalization gives us a way to directly control the temperature (i.e. the scale in my paper) of softmax loss, while we cannot control it through traditional softmax loss because the temperature can be merged into previous layers. 

But feature normalization has many drawbacks. For example, after feature normalization, features can only go along the surface of hypersphere, cannot pass inside. Feature normalization is also unstable near the zero point. A small disturb may cause a feature across the zero point. After normalization, crossing the zero point means getting to the other side of the hypersphere, which is a big change.

**Comment by KaleidoZhouYN** [comment]
@happynear  How about the accuracy on LFW?higher or lower?

**Comment by KaleidoZhouYN** [comment]
@happynear I do think sphereface will be better if combine with sphereface.We can talk about this on WeChat,my WeChat is zyn1000010412

**Comment by happynear** [comment]
I usually don't use WeChat. Maybe you can join our QQ group 347185749.
